doc/Makefile

```diff
@@ -1,63 +1,89 @@
-# Devel: Run `make run` from doc/ to get a live preview of your changes
-# Test:  Run `make spelling` from doc/ to validate your spelling
 #
 
 # You can set these variables from the command line, and also
 # from the environment for the first two.
-SPHINXOPTS    ?=
+SPHINXOPTS    ?= -c . -d .sphinx/.doctrees
 SPHINXBUILD   ?= sphinx-build
+SPHINXDIR     = .sphinx
 SOURCEDIR     = .
 BUILDDIR      = _build
+VENVDIR       = $(SPHINXDIR)/venv
+VENV          = $(VENVDIR)/bin/activate
 
-VENV = env/bin/activate
-PORT = 8090
+.PHONY: help woke-install install run html epub serve clean clean-doc \
+        spelling linkcheck woke Makefile
 
 # Put it first so that "make" without argument is like "make help".
-help:
-	@$(SPHINXBUILD) -M help "$(SOURCEDIR)" "$(BUILDDIR)" $(SPHINXOPTS) $(O)
+help: $(VENVDIR)
+	@. $(VENV); $(SPHINXBUILD) -M help "$(SOURCEDIR)" "$(BUILDDIR)" $(SPHINXOPTS) $(O)
+
+# Explicit target avoids fall-through to the "Makefile" target.
+$(SPHINXDIR)/requirements.txt:
+	test -f $(SPHINXDIR)/requirements.txt
 
-install:
+# If requirements are updated, venv should be rebuilt and timestamped.
+$(VENVDIR): $(SPHINXDIR)/requirements.txt
 	@echo "... setting up virtualenv"
-	python3 -m venv env
-	. $(VENV); pip install --upgrade -r requirements.txt
+	python3 -m venv $(VENVDIR)
+	. $(VENV); pip install --require-virtualenv \
+	    --upgrade -r $(SPHINXDIR)/requirements.txt \
+            --log $(VENVDIR)/pip_install.log
+	@test ! -f $(VENVDIR)/pip_list.txt || \
+            mv $(VENVDIR)/pip_list.txt $(VENVDIR)/pip_list.txt.bak
+	@. $(VENV); pip list --local --format=freeze > $(VENVDIR)/pip_list.txt
 	@echo "\n" \
-	  "--------------------------------------------------------------- \n" \
-      "* watch, build and serve the documentation: make run \n" \
-	  "* check spelling: make spelling \n" \
-	  "\n" \
-      "enchant must be installed in order for pyenchant (and therefore \n" \
-	  "spelling checks) to work. \n" \
-	  "--------------------------------------------------------------- \n"
+        "--------------------------------------------------------------- \n" \
+        "* watch, build and serve the documentation: make run \n" \
+        "* only build: make html \n" \
+        "* only serve: make serve \n" \
+        "* clean built doc files: make clean-doc \n" \
+        "* clean full environment: make clean \n" \
+        "* check links: make linkcheck \n" \
+        "* check spelling: make spelling \n" \
+        "* check inclusive language: make woke \n" \
+        "* other possible targets: make <press TAB twice> \n" \
+        "--------------------------------------------------------------- \n"
+	@touch $(VENVDIR)
+
+woke-install:
+	@type woke >/dev/null 2>&1 || \
+            { echo "Installing \"woke\" snap... \n"; sudo snap install woke; }
+
+install: $(VENVDIR) woke-install
+
+run: install
+	. $(VENV); sphinx-autobuild -b dirhtml "$(SOURCEDIR)" "$(BUILDDIR)" $(SPHINXOPTS)
```

Comment:
`make run` is looping forever for me due to files (being cached?) changes in `.sphinx/breathe/doxygen/`. Adding `--re-ignore '\.sphinx/breathe/doxygen/.*'` to `$SPHINXOPTS` fixed the problem for me.

Reply:
Oops, that option (`--re-ignore`) isn't valid for `sphinx-build`. I'll need to add it separately, not to `$SPHINXOPTS`.